File: 14-cache/npm-package/cache-stats.ts
/**
 * Cost Katana Cache: Monitor Cache Statistics
 * 
 * Track cache performance and cost savings.
 * 
 * Run: npx ts-node 18-cache/npm-package/cache-stats.ts
 */

import axios from 'axios';
import CostKatana from 'cost-katana';

const API_BASE = 'https://cost-katana-backend.store/api';
const API_KEY = process.env.COST_KATANA_API_KEY;

const client = new CostKatana({ apiKey: API_KEY! });

async function getCacheStats() {
  try {
    const response = await axios.get(
      `${API_BASE}/cache/stats`,
      {
        headers: { 'Authorization': `Bearer ${API_KEY}` }
      }
    );

    const stats = response.data.data;

    console.log('📊 Cache Statistics:\n');
    console.log('Redis Cache:');
    console.log(`  Hits: ${stats.redis.hits}`);
    console.log(`  Misses: ${stats.redis.misses}`);
    console.log(`  Total Requests: ${stats.redis.totalRequests}`);
    console.log(`  Hit Rate: ${stats.redis.hitRate.toFixed(2)}%`);
    console.log(`  Cost Saved: $${stats.redis.costSaved.toFixed(2)}`);
    console.log(`  Tokens Saved: ${stats.redis.tokensSaved.toLocaleString()}`);
    
    console.log('\nCombined Statistics:');
    console.log(`  Total Hit Rate: ${stats.combined.hitRate.toFixed(2)}%`);
    console.log(`  Total Cost Saved: $${stats.combined.totalCostSaved.toFixed(2)}`);
    console.log(`  Avg Cost Saved/Request: $${stats.combined.averageCostSaved.toFixed(4)}`);

    return stats;
  } catch (error: any) {
    console.error('❌ Failed:', error.response?.data || error.message);
    throw error;
  }
}

async function demonstrateCaching() {
  console.log('\n🥷 Cache Demonstration\n');

  // First request (cache miss)
  console.log('1️⃣ First request (cache miss)...');
  const response1 = await client.chat.completions.create({
    model: 'gpt-4',
    messages: [{ role: 'user', content: 'What is the capital of France?' }]
  }, {
    headers: {
      'X-Enable-Cache': 'true',
      'X-Cache-TTL': '3600'
    }
  });

  console.log(`  Cache Status: ${response1.headers?.['x-cache-status']}`);
  console.log(`  Response: ${response1.choices[0].message.content?.substring(0, 50)}...`);

  // Second request (cache hit)
  console.log('\n2️⃣ Second request (cache hit)...');
  const response2 = await client.chat.completions.create({
    model: 'gpt-4',
    messages: [{ role: 'user', content: 'What is the capital of France?' }]
  }, {
    headers: { 'X-Enable-Cache': 'true' }
  });

  console.log(`  Cache Status: ${response2.headers?.['x-cache-status']}`);
  console.log(`  Cost Saved: $${response2.headers?.['x-cost-saved']}`);
  console.log(`  Tokens Saved: ${response2.headers?.['x-tokens-saved']}`);

  // Semantic match (similar query)
  console.log('\n3️⃣ Semantically similar request...');
  const response3 = await client.chat.completions.create({
    model: 'gpt-4',
    messages: [{ role: 'user', content: 'Tell me the capital city of France' }]
  }, {
    headers: { 'X-Enable-Cache': 'true' }
  });

  console.log(`  Cache Status: ${response3.headers?.['x-cache-status']}`);
  console.log(`  Similarity Score: ${response3.headers?.['x-similarity-score']}`);
}

async function main() {
  if (!API_KEY) {
    console.error('❌ COST_KATANA_API_KEY required');
    process.exit(1);
  }

  try {
    await getCacheStats();
    await demonstrateCaching();
    
    console.log('\n✅ Cache statistics demo complete!');
    console.log('\n💡 Benefits:');
    console.log('  • 70-90% cache hit rate typical');
    console.log('  • 30-40% cost reduction');
    console.log('  • Instant responses from cache');
    console.log('  • Semantic similarity matching');
  } catch (error) {
    console.error('\n❌ Failed:', error);
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

export { getCacheStats, demonstrateCaching };

File: 14-cache/python-sdk/cache_stats.py
"""
Cost Katana Cache: Monitor Cache Statistics (Python)

Track cache performance and cost savings.

Run: python 18-cache/python-sdk/cache_stats.py
"""

import os
import requests

API_BASE = "https://api.costkatana.com/api"
GATEWAY_URL = f"{API_BASE}/gateway/v1/chat/completions"
API_KEY = os.getenv("COST_KATANA_API_KEY")

def get_cache_stats():
    """Get cache statistics"""
    
    response = requests.get(
        f"{API_BASE}/cache/stats",
        headers={"Authorization": f"Bearer {API_KEY}"}
    )
    
    response.raise_for_status()
    stats = response.json()["data"]
    
    print("📊 Cache Statistics:\n")
    print("Redis Cache:")
    print(f"  Hits: {stats['redis']['hits']}")
    print(f"  Misses: {stats['redis']['misses']}")
    print(f"  Total Requests: {stats['redis']['totalRequests']}")
    print(f"  Hit Rate: {stats['redis']['hitRate']:.2f}%")
    print(f"  Cost Saved: ${stats['redis']['costSaved']:.2f}")
    print(f"  Tokens Saved: {stats['redis']['tokensSaved']:,}")
    
    print("\nCombined Statistics:")
    print(f"  Total Hit Rate: {stats['combined']['hitRate']:.2f}%")
    print(f"  Total Cost Saved: ${stats['combined']['totalCostSaved']:.2f}")
    
    return stats

def demonstrate_caching():
    """Demonstrate caching with API calls"""
    
    print("\n🥷 Cache Demonstration\n")
    
    # First request (cache miss)
    print("1️⃣ First request (cache miss)...")
    r1 = requests.post(
        GATEWAY_URL,
        json={
            "model": "gpt-4",
            "messages": [{"role": "user", "content": "What is the capital of France?"}]
        },
        headers={
            "Authorization": f"Bearer {API_KEY}",
            "X-Enable-Cache": "true",
            "X-Cache-TTL": "3600"
        }
    )
    r1.raise_for_status()
    print(f"  Cache Status: {r1.headers.get('X-Cache-Status')}")
    
    # Second request (cache hit)
    print("\n2️⃣ Second request (cache hit)...")
    r2 = requests.post(
        GATEWAY_URL,
        json={
            "model": "gpt-4",
            "messages": [{"role": "user", "content": "What is the capital of France?"}]
        },
        headers={
            "Authorization": f"Bearer {API_KEY}",
            "X-Enable-Cache": "true"
        }
    )
    r2.raise_for_status()
    print(f"  Cache Status: {r2.headers.get('X-Cache-Status')}")
    print(f"  Cost Saved: ${r2.headers.get('X-Cost-Saved')}")
    print(f"  Tokens Saved: {r2.headers.get('X-Tokens-Saved')}")
    
    # Semantic match
    print("\n3️⃣ Semantically similar request...")
    r3 = requests.post(
        GATEWAY_URL,
        json={
            "model": "gpt-4",
            "messages": [{"role": "user", "content": "Tell me the capital city of France"}]
        },
        headers={
            "Authorization": f"Bearer {API_KEY}",
            "X-Enable-Cache": "true"
        }
    )
    r3.raise_for_status()
    print(f"  Cache Status: {r3.headers.get('X-Cache-Status')}")
    print(f"  Similarity Score: {r3.headers.get('X-Similarity-Score')}")

def main():
    if not API_KEY:
        print("❌ COST_KATANA_API_KEY required")
        return
    
    try:
        get_cache_stats()
        demonstrate_caching()
        
        print("\n✅ Cache statistics demo complete!")
        print("\n💡 Benefits:")
        print("  • 70-90% cache hit rate")
        print("  • 30-40% cost reduction")
        print("  • Instant responses")
        print("  • Semantic matching")
        
    except requests.exceptions.HTTPError as e:
        print(f"❌ Error: {e.response.json() if e.response else e}")

if __name__ == "__main__":
    main()
